service/matching/taskListManager.go

```diff
@@ -474,22 +578,55 @@ func (c *taskListManagerImpl) trySyncMatch(task *persistence.TaskInfo) (*persist
 	}
 	// Request from the point of view of Add(Activity|Decision)Task operation.
 	// But it is getTask result from the point of view of a poll operation.
-	request := &getTaskResult{task: task, C: make(chan *syncMatchResponse, 1)}
+	request := &getTaskResult{task: task, C: make(chan *syncMatchResponse, 1), syncMatch: true}
+
+	rsv := c.rateLimiter.Reserve()
+	if !rsv.OK() {
+		scope := metrics.MatchingTaskListMgrScope
+		c.metricsClient.IncCounter(scope, metrics.AddThrottleCounter)
+		return nil, fmt.Errorf("cannot add to tasklist, limit exceeded")
+	}
 	select {
-	case c.syncMatch <- request: // poller goroutine picked up the task
+	case c.tasksForPoll <- request: // poller goroutine picked up the task
 		r := <-request.C
 		return r.response, r.err
 	default: // no poller waiting for tasks
+		rsv.Cancel()
 		return nil, nil
 	}
 }
 
+func (c *taskListManagerImpl) deliverBufferTasksForPoll() {
+deliverBufferTasksLoop:
+	for {
+		err := c.rateLimiter.Wait(c.cancelCtx)
+		if err != nil {
+			if err == context.Canceled {
+				c.logger.Warn("Tasklist manager context is cancelled, shutting down")
+				break deliverBufferTasksLoop
+			}
+			c.logger.Warn("Unable to send tasks for poll, limit exceeded")
+			c.metricsClient.IncCounter(metrics.MatchingTaskListMgrScope, metrics.AddThrottleCounter)
+			continue
+		}
+		select {
+		case task, ok := <-c.taskBuffer:
+			if !ok { // Task list getTasks pump is shutdown
+				break deliverBufferTasksLoop
+			}
+			c.tasksForPoll <- &getTaskResult{task: task}
```

Comment:
We should measure the perf impact of this change.  Considering we are not serializing all tasks through a non-buffered channel.  Can you try to add a bench test to cover this scenario?

Reply:
Are you talking about preference for buffered tasks over sync tasks? We discussed that and came to the conclusion that even before this, we will still  prefer buffer tasks over sync tasks.

Reply:
I think it is going to prioritize sync match just because sync match uses multiple goroutines to call trySyncMatch while deliverBufferTasksToPoll is called from a single goroutine.

Reply:
Sync match is not in multiple goroutines. Do you mean there are several parallel calls to AddTask whereas only one goroutine per host adding the buffer tasks? That is true, and that behavior would be true before this code change as well.